File: public/data/chapters/chapter_01_arithmetic/logic_advanced.py
"""
Advanced Arithmetic Logic for Python Runtime.
Supports add, mul, sin, cos.
"""
import math

def calculate(target_content):
    """
    Compute based on 'op' in context.
    """
    # 'context' is injected into the global scope by the runtime
    global context
    
    try:
        if context.get("batch"):
            results = []
            for ex in context.get("examples", []):
                op = ex.get("op", "add")
                a = ex.get("a", 0)
                b = ex.get("b", 0)
                res = 0
                if op == "add": res = a + b
                elif op == "mul": res = a * b
                elif op == "sin": res = math.sin(a)
                elif op == "cos": res = math.cos(a)
                results.append(res)
            return results

        op = context.get("op", "add")
        a = context.get("a", 0)
        b = context.get("b", 0)
        
        if op == "add":
            return a + b
        elif op == "mul":
            return a * b
        elif op == "sin":
            return math.sin(a)
        elif op == "cos":
            return math.cos(a)
        return 0
    except Exception as e:
        return f"Error in Python: {str(e)}"

File: public/data/chapters/chapter_01_arithmetic/logic_advanced.js


// Read context from input
// The runtime wrapper injects 'input' variable containing {target, context}

if (context.batch) {
    result = (context.examples || []).map(ex => {
        const op = ex.op || "add";
        const a = ex.a || 0;
        const b = ex.b || 0;
        if (op === "add") return a + b;
        if (op === "mul") return a * b;
        if (op === "sin") return Math.sin(a);
        if (op === "cos") return Math.cos(a);
        return 0;
    });
} else {
    // Single execution
    const op = context.op || "add";
    const a = context.a || 0;
    const b = context.b || 0;

    if (op === "add") {
        result = a + b;
    } else if (op === "mul") {
        result = a * b;
    } else if (op === "sin") {
        result = Math.sin(a);
    } else if (op === "cos") {
        result = Math.cos(a);
    }
}

// Result is captured by the wrapper
